# There are Many APIs
- There are organizations tyhat put up public APIs and sell access to those APIs
- We will explore a geocoding API based on the OpenStreetMap data
You need an account to access this API
- There is a free level of requests
- You pay above that rate of usage
- https://geoapify.com/

## An API Proxy
- To avoid making you get an account, I have a well-hidden web server that acts as a proxy for the Geoapify data
- This proxy does not require a password - but it does have rate limits and is heavily cached using an edge-caching service for performance

Your code <-> cloudfare <-> py4e data <-> Geoapi

In [3]:
import urllib.request, urllib.parse, urllib.error
import xml.etree.ElementTree as ET
import ssl
import json

api_key = False
# If you have a Google Places API key, enter it here
# api_key = 'AIzaSy___IDByT70'
# https://developers.google.com/maps/documentation/geocoding/intro

if api_key is False:
    api_key = 42
    serviceurl = 'http://py4e-data.dr-chuck.net/xml?'
else :
    serviceurl = 'https://maps.googleapis.com/maps/api/geocode/xml?'

# Ignore SSL certificate errors
ctx = ssl.create_default_context()
ctx.check_hostname = False
ctx.verify_mode = ssl.CERT_NONE

while True:
    address = input('Enter location: ')
    if len(address) <  1:
        break
    address = address.strip()
    parms = dict()
    parms['q'] = address
    
    url = serviceurl + urllib.parse.urlencode(parms)
    
    print('Retrieving', url)
    uh = urllib.request.urlopen(url, context=ctx)
    data = uh.read().decode()
    print('Retrieved', len(data), 'characters', data[:20].replace('\n', ' '))
    
    js = json.loads(data)
    
    lat = js['features'][0]['properties']['lat']
    lon = js['features'][0]['properties']['lon']
    print('lat', lat, 'lon', lon)
    location = js['features'][0]['properties']['formatted']
    print(location)

Retrieving https://py4e-data.dr-chuck.net/opengeo?q=Ann+Arbor%2C+MI
Retrieved 1318 characters {"type":"FeatureColl
lat 42.2813722 lon -83.7484616
Ann Arbor, MI, United States of America
Retrieving https://py4e-data.dr-chuck.net/opengeo?q=Ann+Arbor%2C+MI
Retrieved 1318 characters {"type":"FeatureColl
lat 42.2813722 lon -83.7484616
Ann Arbor, MI, United States of America


Retrieving https://py4e-data.dr-chuck.net/opengeo?q=Ann+Arbor%2C+MI

Retrieved 1318 characters {"type":"FeatureColl

lat 42.2813722 lon -83.7484616

Ann Arbor, MI, United States of America

# Summary
- Service Oriented Architecture - allows an application to be brokem into parts and distributed across a network
- An Application Program Interface (API) is a contract for interaction
- Web Services provide infrastructure for applications cooperating (an API) over a network - SOAP and REST are two styles of web services
- XML and JSON are serialization formats